File: src/Sort/InsertionSort.hs
-- MIT License

-- Copyright (c) 2017 Oskar Mendel

-- Permission is hereby granted, free of charge, to any person obtaining a copy
-- of this software and associated documentation files (the "Software"), to deal
-- in the Software without restriction, including without limitation the rights
-- to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-- copies of the Software, and to permit persons to whom the Software is
-- furnished to do so, subject to the following conditions:

-- The above copyright notice and this permission notice shall be included in all
-- copies or substantial portions of the Software.

-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.

-- Recursively iterates over a list to put an element in its correct spot.
insert :: (Ord a) => a -> [a] -> [a]
insert x [] = [x]
insert x (y:ys)
  | x > y = y : insert x ys
  | otherwise = x : y : ys

-- Insertion sort iterates, consuming one input element each repetition,
-- and growing a sorted output list. Each iteration, insertion sort
-- removes one element from the input data, finds the location it belongs
-- within the sorted list, and inserts it there. It repeats until no input
-- elements remain.
--
-- Worst-case - O(n^2)
insertionSort :: (Ord a) => [a] -> [a]
insertionSort x = foldr insert [] x

File: src/Sort/InsertionSort.cpp
//
// Created by Oskar Mendel on 4/23/17.
//

#include "InsertionSort.h"

/**
 * Insertion sort iterates, consuming one input element each repetition,
 * and growing a sorted output list. Each iteration, insertion sort
 * removes one element from the input data, finds the location it belongs
 * within the sorted list, and inserts it there. It repeats until no input
 * elements remain.
 *
 * Worst-case - O(n^2)
 *
 * @param values - Array with values to be sorted.
 * @param size - Size of array with values.
 */
void InsertionSort(int values[], int size) {
    for (int i = 1; i < size; i++) {
        int key = values[i];
        int j = i-1;

        /* Move elements of arr[0..i-1], that are
         greater than key, to one position ahead
         of their current position */
        while (j >= 0 && values[j] > key)
        {
            values[j+1] = values[j];
            j = j-1;
        }
        values[j+1] = key;
    }
}